lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -0,0 +1,257 @@
+
+        return dashboards
+
+    def _encode_dashboard_content(self, content: Union[str, bytes]) -> str:
+        if isinstance(content, str):
+            content = bytes(content, "utf-8")
+
+        return base64.b64encode(lzma.compress(content)).decode("utf-8")
+
+
+class COSMachineDataChanged(EventBase):
+    """Event emitted when a `COSMachineProvider` joins or updates data."""
+
+
+class COSMachineConsumerEvents(ObjectEvents):
+    """Event descriptor for events raised by `COSMachineConsumer`."""
+
+    data_changed = EventSource(COSMachineDataChanged)
+
+
+class COSMachineConsumer(Object):
+    """Provider class for the cos_machine interface."""
+
+    on = COSMachineConsumerEvents()
+
+    def __init__(
+        self,
+        charm,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        refresh_events: Optional[List] = None,
+    ):
+        """Create a COSMachineConsumer instance.
+
+        Args:
+            charm: The `CharmBase` instance that is instantiating this object.
+            relation_name: The name of the relation to communicate over.
+            refresh_events: List of events on which to resfresh relation data.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._refresh_events = refresh_events or [self._charm.on.config_changed]
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self._on_relation_data_changed)
+        self.framework.observe(events.relation_changed, self._on_relation_data_changed)
+        for event in self._refresh_events:
+            self.framework.observe(event, self.trigger_refresh)
+
+    def _on_relation_data_changed(self, _):
+        self.on.data_changed.emit()
+
+    def trigger_refresh(self, _):
+        """Trigger a refresh of relation data."""
+        # FIXME: Figure out what we should do here
+        self.on.data_changed.emit()
+
+    @property
+    def _relations(self):
+        return self._charm.model.relations[self._relation_name]
+
+    @property
+    def metrics_jobs(self) -> List[Dict]:
+        """Return a prometheus_scrape-like data structure for jobs."""
+        scrape_jobs = []
+        for relation in self._relations:
+            config = json.loads(relation.data[relation.app].get("config", {}))
+            if jobs := config.get("metrics", {}).get("scrape_jobs", []):
+                for job in jobs:
+                    job_config = {
+                        "job_name": job["job_name"],
+                        "metrics_path": job["path"],
+                        "static_configs": [{"targets": [f"localhost:{job['port']}"]}],
+                    }
+                    scrape_jobs.append(job_config)
+
+        return scrape_jobs
+
+    @property
+    def metrics_alerts(self) -> Dict[str, Any]:
+        """Fetch metrics alerts."""
+        alert_rules = {}
+        for relation in self._relations:
```

Comment:
same as above: try/except ModelError in case relation data is dead.
You could abstract away the yield-relation-data part of this method and the others in a separate `_yield_configs` method to make it less verbose.